src/atomic/atomic_cell.rs

```diff
@@ -737,8 +784,26 @@ where
             mem::transmute_copy(&a.load(Ordering::SeqCst))
         },
         {
-            let _lock = lock(src as usize);
-            ptr::read(src)
+            let lock = lock(src as usize);
+
+            // Try doing an optimistic read first.
+            if let Some(stamp) = lock.optimistic_read() {
+                // We need a volatile read here because other threads might concurrently modify the
```

Comment:
From an extremely pedantic view (which I don't agree that much), volatile read is still UB in C/C++.  But everyone is using volatile read/write to mark an access concurrent, including Linux for starters.  Maybe commenting on it would be helpful to readers or ASAN?

Reply:
I added a more elaborate comment explaining the issue.